src/care/host_device_map.h

```diff
@@ -0,0 +1,189 @@
+#define _CARE_DEVICE_UNORDERED_MAP_H
+#include "care/config.h"
+#include "care/atomic.h"
+
+#include <unordered_map>
+#include <care/KeyValueSorter.h>
+
+namespace care {
+
+   ///
+   /// @author Peter Robinson
+   ///
+   /// @class host_device_map is a rudimentary associative map. On the host,
+   /// it is backed by a std::unordered_map, and on the device it uses a
+   /// care::KeyValueSorter.
+   ///
+   /// Restrictions when compared to std::map:
+   ///   1. Total capacity must be declared up front.
+   ///   2. a call to sort() must be performed before any lookups are done.
+   ///   3. Insertions can only be done via an emplace(Key,Value).
+   ///   4. Lookups can only be done via an at(Key)
+   ///
+   template <typename key_type,
+             typename mapped_type,
+             typename Exec>
+   class host_device_map
+   {
+      public:
+         host_device_map(size_t max_entries);
+         CARE_HOST_DEVICE inline void emplace(key_type key, mapped_type val) const;
+         CARE_HOST_DEVICE inline mapped_type at(key_type key) const;
+         void sort();
+         void free();
+   };
+
+// ********************************************************************************
+// RAJA::seq_exec specialization.
+// ********************************************************************************
+   template <typename key_type, typename mapped_type>
+   class host_device_map< key_type, mapped_type, RAJA::seq_exec> {
+      public:
+         // constructor
+         host_device_map(size_t max_entries) : m_map()  {
+            m_map = new std::unordered_map<key_type, mapped_type>{};
+         }
+        // emplace a key value pair
+        inline void emplace(key_type key, mapped_type val) const {
+           (*m_map)[key] = val;
+        }
+        // lookup a value
+        inline mapped_type at(key_type key) const {
+           return (*m_map).at(key);
+        }
+        // prepare for lookups (no-op for std::unordered map implementation)
+        void sort() { }
+
+        // delete the map
+        void free() { delete m_map; }
+
+      private:
+         // we do a heap allocated map to ensure no deep copies occur during lambda capture
+         std::unordered_map<key_type, mapped_type> * m_map;
+   };
+
+
+#ifdef CARE_GPUCC
+
+// ********************************************************************************
+// RAJADeviceExec specialization.
+// ********************************************************************************
+   template <typename key_type, typename mapped_type>
+   class host_device_map<key_type, mapped_type, RAJADeviceExec>
+   {
+      public:
+         using int_ptr = care::host_device_ptr<int>;
+         // constructor
+         host_device_map(size_t max_entries)  {
+            // m_size will be atomically incremented as elements are emplaced into the map
+            m_size = int_ptr(1, "map_size");
+            int_ptr size = m_size;
+            CARE_SEQUENTIAL_LOOP(i, 0, 1) {
```

Comment:
Use CARE_KERNEL to avoid `i` being unused.

Reply:
fixed